### PR TITLE
Provide File Copier plugin

### PR DESCRIPTION
**Pull request by RemiGuillard**

### CODE REVIEW

**Comment by kamaradclimber** [comment]
I cfannot review this on my phone, can you wait until monday or find
another reviewer?

File: README.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
 Rundeck WinRM Plugin
 --------------------
 
-This is a [Rundeck Node Execution plugin][1] that uses WinRM to connect to Windows and execute commands.  It uses the [OverThere Library][2] to provide the WinRM implementation, and uses Basic authentication over HTTPS.
+These are a [Rundeck Node Execution plugin and a Rundeck File Copier plugin][1] that uses WinRM to connect to Windows and execute commands.  It uses the [OverThere Library][2] to provide the WinRM implementation, and uses Basic authentication over HTTPS.
 
-[1]: http://rundeck.org/docs/manual/plugins.html#node-execution-plugins
+[1]: http://rundeck.org/2.3.2/plugins-user-guide/node-execution-plugins.html
```
Comment (BenoitPerrot) [inline README.md]: Keep links unrelated to the version of Rundeck, as this was visibly the intent of the text before your patch (see comment above for the appropriate links)

File: README.md [hunk 2]
```diff
@@ -28,18 +28,22 @@ If using Kerberos for authentication, also follow these guides:
...
+	<node name="winNode" file-copier="overthere-winrm" .../>
+
+Or set as the default NodeExecutor for your project/framework properties file, with `service.NodeExecutor.default.provider=overthere-winrm` and `service.FileCopier.default.provider=overthere-winrm`.
 
 These Node attributes are used to connect to the remote host:
 
 * `username` - Remote username. If using Kerberos, see [Using Kerberos Authentication](#using-kerberos-authentication)
 * `hostname` - Remote host. Can include "host:port" to specify port number other than the default 5985/5986 (http/https).
 
-Password authentcation can be performed in one of two ways:
```
Comment (BenoitPerrot) [inline README.md:42]: This change is unrelated. If you absolutely want to push it, move it to another patch (eg "README: fix typos").

File: README.md [hunk 3]
```diff
@@ -1,9 +1,9 @@
 Rundeck WinRM Plugin
 --------------------
 
-This is a [Rundeck Node Execution plugin][1] that uses WinRM to connect to Windows and execute commands.  It uses the [OverThere Library][2] to provide the WinRM implementation, and uses Basic authentication over HTTPS.
+These are a [Rundeck Node Execution plugin and a Rundeck File Copier plugin][1] that uses WinRM to connect to Windows and execute commands.  It uses the [OverThere Library][2] to provide the WinRM implementation, and uses Basic authentication over HTTPS.
```
Comment (BenoitPerrot) [inline README.md]: Split this in two links, are there are two different pages:
- http://rundeck.org/docs/plugins-user-guide/node-execution-plugins.html
- http://rundeck.org/docs/developer/file-copier-plugin.html

**Review by BenoitPerrot — CHANGES_REQUESTED** [review]

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+                    dst = connection.getFile(destination);
+                    context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "Destination file acquired " + dst.getPath());
+                } else {
+                    dst = connection.getTempFile("RD_WINRM_TMP");
+                    context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "Temporary file acquired " + dst.getPath());
+                }
+                context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "Destination file acquired " + dst.getPath());
+
+                try {
+                    OutputStream w = dst.getOutputStream();
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: There is a dedicated method in IOUtils for that: `copy(InputStream input, OutputStream output)`

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 5]
```diff
@@ -0,0 +1,805 @@
...
+ * alter the password resolution mechanism.
+ * See {@link #getKerberosUsername(com.dtolabs.rundeck.plugin.overthere.OTWinRMFileCopier.ConnectionOptionsBuilder)}
+ * and
+ * {@link #getClearAuthPassword(com.dtolabs.rundeck.plugin.overthere.OTWinRMFileCopier.ConnectionOptionsBuilder)}.
+ *
+ * The connection options can be altered as well by overriding {@link #willUseConnectionOptions(ConnectionOptions)}.
+ */
+@Plugin(name = OTWinRMFileCopier.SERVICE_PROVIDER_TYPE, service = "FileCopier")
+public class OTWinRMFileCopier implements DestinationFileCopier, Describable {
+    public static final String SERVICE_PROVIDER_TYPE = "overthere-winrm";
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: The first ~150 lines of this file are exactly the same as in OTWinRMNodeExecutor.java. Move them to a dedicated class,  containing only static infos, and use it in both classes

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 6]
```diff
@@ -0,0 +1,805 @@
...
+                    context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "Temporary file acquired " + dst.getPath());
+                }
+                context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "Destination file acquired " + dst.getPath());
+
+                try {
+                    OutputStream w = dst.getOutputStream();
+                    w.write(IOUtils.toByteArray(inputStream));
+                    w.close();
+                } catch (IOException ioe) {
+                    String message = null;
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: This block is duplicated at lines 269 and 279. Factor.

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 7]
```diff
@@ -0,0 +1,805 @@
...
+        return DESC;
+    }
+
+    @Override
+    public String copyFileStream(ExecutionContext context, InputStream inputStream, INodeEntry node, String destination) throws FileCopierException {
+        ConnectionOptions options = null;
+        String logprompt = "[" + SERVICE_PROVIDER_TYPE + ":" + node.extractHostname() + "] ";
+
+        if (null == context.getExecutionListener()) {
+            System.out.println(logprompt + " Bad plugin context!  NULL ExecutionListener");
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java:69]: .log() instead of println()

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 8]
```diff
@@ -0,0 +1,805 @@
...
+                        ioe.printStackTrace(System.err);
+                        message = ioe.getMessage();
+                    } else {
+                        message = "WinRM Error: " + ioe.getMessage();
+                    }
+                    context.getExecutionListener().log(Constants.ERR_LEVEL, logprompt + "failed: " + message);
+                    throw new FileCopierException(message, Reason.IOWriteError);
+                }
+                context.getExecutionListener().log(Constants.VERBOSE_LEVEL, logprompt + "File copied at: " + dst.getPath());
+                System.out.println(dst.getPath());
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Remove, or use .log() instead of println()

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 9]
```diff
@@ -0,0 +1,805 @@
...
+    }
+
+    @Override
+    public String copyFile(ExecutionContext context, File file, INodeEntry node, String destination) throws FileCopierException {
+
+        String filePath;
+
+        try {
+            InputStream fileStream = new FileInputStream(file);
+            filePath = copyFileStream(context, fileStream, node, destination);
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Remove filePath: just go for `return copyFileStream...`

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 10]
```diff
@@ -0,0 +1,805 @@
...
+        return filePath;
+    }
+
+    @Override
+    public String copyScriptContent(ExecutionContext context, String script, INodeEntry node, String destination) throws FileCopierException {
+
+        String filePath;
+
+        InputStream scriptStream = new ByteArrayInputStream(script.getBytes());
+        filePath = copyFileStream(context, scriptStream, node, destination);
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Remove filePath: just go for `return copyFileStream...`

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 11]
```diff
@@ -0,0 +1,805 @@
...
+    }
+
+    @Override
+    public String copyFile(ExecutionContext context, File file, INodeEntry node) throws FileCopierException {
+
+        String filePath;
+
+        try {
+            InputStream fileStream = new FileInputStream(file);
+            filePath = copyFileStream(context, fileStream, node, null);
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Remove filePath: just go for `return copyFileStream...`

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 12]
```diff
@@ -0,0 +1,805 @@
...
+        return filePath;
+    }
+
+    @Override
+    public String copyScriptContent(ExecutionContext context, String script, INodeEntry node) throws FileCopierException {
+
+        String filePath;
+
+        InputStream scriptStream = new ByteArrayInputStream(script.getBytes());
+        filePath = copyFileStream(context, scriptStream, node, null);
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Remove filePath: just go for `return copyFileStream...`

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 13]
```diff
@@ -0,0 +1,805 @@
...
+    public static enum Reason implements FailureReason {
+        WinRMProtocolError,
+        FileNotStreamableError,
+        IOWriteError,
+    }
+
+
+
+
+    /**
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: A large part of the code below (all of it?) is the same as in the NodeExecutor.
Extract it in a dedicated class, and use it. It might be the same as the one I was suggesting at the beginning of the file.

**Review by BenoitPerrot — CHANGES_REQUESTED** [review]

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java (hunk 7 above)
Comment (RemiGuillard) [reply src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java:69]: Can not here as the log() is precisely in the context.getExecutionListener()

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java (hunk 8 above)
Comment (RemiGuillard) [reply src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: Removed

File: README.md (hunk 3 above)
Comment (RemiGuillard) [reply README.md]: These are not the equivalents links. FileCopier and NodeExecutor are "node execution" plugin category
But agree to add the link to the considered plugin

File: README.md (hunk 1 above)
Comment (RemiGuillard) [reply README.md]: Agreed

File: README.md (hunk 2 above)
Comment (RemiGuillard) [reply README.md:42]: Hagrid. Done in the following commit

**Review by BenoitPerrot — APPROVED** [review]

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+                    "Destination folder for remote scripts", false, null))
+            .mapping(CONFIG_DST_FILE, PROJ_PROP_PREFIX + FILE_COPY_DESTINATION_DIR)
+            .build();
+
+    public Description getDescription() {
+        return DESC;
+    }
+
+
+    public String buildErrorMEssage(ExecutionContext context, Exception ex, String logprompt) {
```
Comment (BenoitPerrot) [inline src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: typo: should be buildErrorMessage instead of buildErrorMEssage

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java (hunk 7 above)
Comment (BenoitPerrot) [reply src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java:69]: yuck. okay.

**Review by BenoitPerrot — COMMENTED** [review]
Please change the commit message. Keep the first line simple, and detail in the body:

"""
OTWinRMNodeExecutor: Extract reusable members into dedicated class

Preparing the introduction of an over-WinRM file copier.
"

File: src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java (hunk 14 above)
Comment (BenoitPerrot) [reply src/main/java/com/dtolabs/rundeck/plugin/overthere/OTWinRMFileCopier.java]: (bis: should be buildErrorMessage instead of buildErrorMEssage)

**Review by BenoitPerrot — CHANGES_REQUESTED** [review]

**Review by BenoitPerrot — APPROVED** [review]
